### PR TITLE
Desktop App: fix crash on Apple Silicon caused by 8.0.0.

### LINKED ISSUE #78884: WordPress Desktop app 8.0.0: crash while trying to open it on Apple Silicon
### Quick summary

When trying to open the last version of the Desktop app(8.0.0) we get a Apple crash report.

The issue was raised in Slack initially here (p1688130305943999-slack-C08GLCXMZ).

### Steps to reproduce

1. Try to open the WordPress desktop app 8.0.0:

https://github.com/Automattic/wp-desktop/releases/tag/v8.0.0

### What you expected to happen

To work as expected

### What actually happened

It crashed instead.

### Impact

Some (< 50%)

### Available workarounds?

Use the 7.2.0 version instead: https://github.com/Automattic/wp-desktop/releases/download/v7.2.0/wordpress.com-macOS-app-7.2.0-arm64.zip

### Platform (Simple and/or Atomic)

_No response_

### Logs or notes

Filed this report on a request from @worldomonation.

### PR DESCRIPTION
**Pull request by worldomonation**
Fixes https://github.com/Automattic/wp-calypso/issues/78884.

## Proposed Changes

This PR adds fixes to the Entitlements list that caused the Apple Silicon build of the app to either crash or display a white screen on startup.

This PR also adds a fix for the failing Window build.

## What went wrong?

The previous release of WordPress.com Desktop App 8.0.0 caused crashes on users on the Apple Silicon architecture.

After almost a week of investigations, debugging and trial I found the culprit is because we jumped from Electron 12.x.x to Electron 24.x.x.

More specifically, [Electron has required (since 20.x.x)](https://github.com/electron/electron/issues/35355#issuecomment-1227574710) that additional entitlement be added in the form of `com.apple.security.cs.allow-jit`. 

This issue was not caught in https://github.com/Automattic/wp-calypso/pull/78098, because:
1. CircleCI builds were all successful;
2. locally built Apple Silicon builds launched as expected, and it was assumed that CI-produced builds will be the same;

As it turns out, the CI-produced builds involve the Apple signing process, which is where the Entitlements come into play. Because the Entitlements didn't change from Electron 12.x.x, when the binary was signed additional privileges required to run Electron >= 20.x.x was not granted, hence the app crashed on all Apple Silicon users.

## Testing Instructions

Ensure the following build configurations are passing:
  - [ ] CircleCI > wp-desktop-mac
  - [ ] CircleCI > wp-desktop-linux
  - [ ] CircleCI > wp-desktop-windows

Additionally, download the x64 and arm64 produced binaries from CircleCI and ensure it functions as expected.

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Review by wp-desktop — DISMISSED** [review]
WordPress Desktop CI Failure for job "wp-desktop-mac".

@worldomonation please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/1122323). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

File: desktop/package.json
Comment (nsakaimbo) [inline desktop/package.json]: `electron-rebuild` should be a dev dependency, and not a direct dependency of the app.

File: .circleci/config.yml
Comment (nsakaimbo) [inline .circleci/config.yml:171]: Any specific reason for using the `m1` image?

**Comment by worldomonation** [comment]
Status update, as of July 6, 2023:

[Build](https://app.circleci.com/pipelines/github/Automattic/wp-calypso/183694/workflows/ffea0b44-ff01-47d3-82ab-250ba9f3c9ba/jobs/1122471/steps)

- resolved the issue with entitlement file structure, the build can now complete.
- E2E tests still can't run, likely related to the issue below.

Configuration:
- electron 25.2.0
- electron/rebuild 3.2.13
- electron-builder 23.0.3

**Issue**
The app loads a white screen of death.
image

File: .circleci/config.yml
Comment (worldomonation) [inline .circleci/config.yml:171]: @nsakaimbo What do you think? Is there a cost penalty incurred to use the M1? 

Performance wise, we see a nice speed-up with the M1 builders. 
Compare the below:

image

The left shows the M1 builder and the right shows the Intel builder.
Even though the M1 builder is building both the x64 and arm64 binaries, the overall execution time for the build is the same as the Intel builder, which is only building one binary.

Should we go back to building only one binary by default, this would mean the overall task execution time being cut down by as much as a minute.

File: desktop/bin/build-mac-ci.js
Comment (worldomonation) [inline desktop/bin/build-mac-ci.js]: I've elected to remove the conditional and thus always build both the Intel and Apple Silicon binaries. 

I am thinking that as more of the world moves to Apple Silicon, making sure the arm64 build doesn't break is going to be important.

Comment (nsakaimbo) [reply desktop/bin/build-mac-ci.js]: > I've elected to remove the conditional and thus always build both the Intel and Apple Silicon binaries.

I would suggest not doing this, and limiting the build to only one or the other (preferably the arm64 build, as we should be deprecating the Intel build soon anyway).

From the CI run ([link](https://app.circleci.com/pipelines/github/Automattic/wp-calypso/183709/workflows/e0cf0e0a-de82-4043-806e-8b1763f1150c/jobs/1122495)), it's unclear which architecture the end-to-end tests were ran against (as far as I can tell it wasn't both). It's nice that the test script can find either the Apple Silicon or Intel binaries, but you still have to specify which of the two builds to use during a test run (or if you really want to build both, then amend the CI script to execute against both):

```
jest-haste-map: Haste module naming collision: WordPressDesktop
  The following files share their name; please adjust your hasteImpl:
    * <rootDir>/release/mac/WordPress.com.app/Contents/Resources/app/package.json
    * <rootDir>/release/mac-arm64/WordPress.com.app/Contents/Resources/app/package.json

 PASS  test/e2e/specs/login.js
  User Can log in
    ✓ Log in (950 ms)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        3.645 s
Ran all test suites.
```

File: .circleci/config.yml
Comment (nsakaimbo) [reply .circleci/config.yml:171]: That's great to see that the M1 image is a lot more performant. 👏🏽  I'm not sure about any cost penalty, though - to my knowledge, we are only able to use images that are accessible with our existing Circle plan, so I think this should be OK?

> Should we go back to building only one binary by default, this would mean the overall task execution time being cut down by as much as a minute.

Yeah I think we should limit the binaries generated during non-release runs to just one (see my other comment on this point).

File: desktop/package.json
Comment (nsakaimbo) [inline desktop/package.json]: Quick questions: Why is the bump to electron version necessary? If we maintain the same dependencies, the changes to the entitlement files should be sufficient for this fix, no?

In general, we should be a little more cautious when bumping major electron versions, as the process needs to be a little more comprehensive. For starters, you would need to evaluate any breaking API changes as described in Electron's [breaking changes docs](https://www.electronjs.org/docs/latest/breaking-changes), and smoke-test the app on all platforms as there are things that definitely won't be caught by CI (for example menu items, keyboard shortcuts, the auto-updater). With the exception of the Apple Silicon build, I did verify these things when updating to version 8.0.0. Unless you're willing and able to verify all these items again (on all platforms), I would suggest maintaining the electron version and other dependencies. The less frequently we have to perform this sort of manual verification/smoke-testing, the better IMO.

File: desktop/package.json
Comment (nsakaimbo) [inline desktop/package.json]: By convention we typically save the version bump for an actual release PR. The release itself should be isolated from any other code-changes.

File: desktop/test/e2e/specs/login.js
Comment (nsakaimbo) [inline desktop/test/e2e/specs/login.js:22]: Nice. 👍🏽 Having this in the code is useful. 

However as I noted in my other comment, it's unclear which architecture CI is using for the e2e run. 🤔 My preference would be to build only one architecture for non-release CI runs, but if building both you would want to make it so the test suite is actually ran against both.

Comment (worldomonation) [reply desktop/test/e2e/specs/login.js:22]: CI will use whatever architecture it detects in `process.arch`.

On Intel builders, it would be the x64 binaries. On Apple Silicon builders, it would be the arm64 binaries.

The issue this addresses, is that previously were were using Intel builders to build x64 and arm64 builds. It worked fine when the primary build we are producing was x64, but as soon as we defaulted to producing arm64 the E2E tests can no longer execute because Intel builders cannot run Apple Silicon binaries.

File: desktop/package.json
Comment (worldomonation) [reply desktop/package.json]: Interesting, I thought a major bugfix like this would require a version bump of some sort.

File: desktop/package.json
Comment (worldomonation) [reply desktop/package.json]: I updated the packages initially as part of our private conversation about using the up-to-date version of electron-related packages.

Though I didn't check all of the breaking changes doc, I did get my teammates to check macOS builds and I checked Windows myself. Is there more in-depth checks required if bumping electron versions?

File: desktop/bin/build-mac-ci.js
Comment (worldomonation) [reply desktop/bin/build-mac-ci.js]: > if you really want to build both, then amend the CI script to execute against both

To clarify, the proposed choices are either trim it back to building and testing a single architecture (like the status quo), or build both and run tests for both binaries. 

Edit: I saw [this comment](https://github.com/Automattic/wp-calypso/pull/78969#discussion_r1258983208) and it sounds like you're in favor of cutting it back to a single arm64 build for non-releases.

If so, I can look at getting this out, but with a high-visibility project today and tomorrow I am not sure how much time I have to work on it.

File: desktop/test/e2e/specs/login.js
Comment (worldomonation) [reply desktop/test/e2e/specs/login.js:22]: One other thing to note is that having this conditional makes it possible for both Intel and Apple Silicon processors to build and run the E2E builds locally.

File: desktop/package.json
Comment (worldomonation) [reply desktop/package.json]: Reverted in 3e0a8ece862d8eb2f51604ff715eea2a7515c195.

File: desktop/package.json
Comment (worldomonation) [reply desktop/package.json]: Reverted in 3e0a8ece862d8eb2f51604ff715eea2a7515c195.

**Review by wp-desktop — DISMISSED** [review]
WordPress Desktop CI Failure for job "wp-desktop-mac".

@worldomonation please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/1122935). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

**Comment by worldomonation** [comment]
@nsakaimbo One of the reverted changes ended up causing the build to break. I'll have to investigate, but can we settle on an acceptable solution to get the fix out for users complaining about the crashes?

Edit: investigated arm64 build on Apple Silicon, we're back to a white screen of death.

Edit Edit: fixed the issue, the core fix for the crash/white screen got dropped when I rebased.

File: desktop/test/e2e/specs/login.js
Comment (nsakaimbo) [reply desktop/test/e2e/specs/login.js:22]: > CI will use whatever architecture it detects in process.arch.

ah, yes - I missed this detail. 👍🏽 

> One other thing to note is that having this conditional makes it possible for both Intel and Apple Silicon processors to build and run the E2E builds locally.

Yep - this is a nice change to the script.

File: desktop/package.json
Comment (nsakaimbo) [reply desktop/package.json]: > Is there more in-depth checks required if bumping electron versions?

I tried to enumerate them in [this preliminary PR](https://github.com/Automattic/wp-calypso/pull/75516#issue-1661538958) from a while ago. Aside from smoke-testing, the auto-updater is a critical component. Included in manual smoke-testing is ensuring keyboard shortcuts and menu items continue to work as expected.

File: desktop/bin/build-mac-ci.js
Comment (nsakaimbo) [reply desktop/bin/build-mac-ci.js]: > it sounds like you're in favor of cutting it back to a single arm64 build for non-releases.

Yes - I would lean towards a single architecture for non-release builds:

```
const arches = isReleaseBuild ? [ 'x64', 'arm64' ] : [ 'arm64' ];
```

I would consider the Intel build less important going forward (perhaps it's something we can eventually drop completely).

**Review by nsakaimbo — APPROVED** [review]
Nice work, and way to go in identifying that entitlements fix. 👍🏽 For a quick sanity check I tested the x64 artifact as generated by the M1 image and confirmed it worked correctly. 🎉 

I would suggest limiting builds to a single architecture for non-release builds, but ultimately that's up to you and LGTM otherwise.

File: desktop/bin/build-mac-ci.js
Comment (worldomonation) [reply desktop/bin/build-mac-ci.js]: Sounds good, will make the change.

**Review by wp-desktop — DISMISSED** [review]
WordPress Desktop CI Failure for job "wp-desktop-mac".

@worldomonation please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/1123098). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

**Comment by worldomonation** [comment]
@nsakaimbo Thanks for the review! This PR is ready to go, with all builds building correctly and passing.

One other thing I had to fix was the artifact path for yet another step: Persist Mac Executable. 

After fixing the path, I noticed the **artifact list isn't as descriptive as it used to be when building two arches**:
image

When building two arches:
image

From comparing the artifact list to other trunk builds, it seems like normally the "less descriptive" artifacts are captured, but I want to make sure this looks alright before merging the fix.


Links
Two arches: https://app.circleci.com/pipelines/github/Automattic/wp-calypso/184425/workflows/b35a8043-c225-4df8-963a-0cc8eb09d75e/jobs/1123016/artifacts
One arch: https://app.circleci.com/pipelines/github/Automattic/wp-calypso/184577/workflows/420582f2-7535-4b74-85a8-b8d8c20edb19/jobs/1123102/artifacts

**Comment by nsakaimbo** [comment]
>  I noticed the artifact list isn't as descriptive as it used to be when building two arches:

Good catch, but yes this is by design. If you look at the existing artifacts being persisted for business-as-usual jobs in `trunk` right now, they would match the smaller artifact set (i.e. the app zip and contents of the `results/` folder). The full artifact set should be persisted only at release time. My guess is there are probably a few more places in the build scripts that would need to accommodate two architectures being more cleanly persisted for non-release jobs. In addition, it seems the SHA of the "two arches" [CI job](https://app.circleci.com/pipelines/github/Automattic/wp-calypso/184425/workflows/b35a8043-c225-4df8-963a-0cc8eb09d75e/jobs/1123016/artifacts) you shared doesn't seem to be available any longer - it's possible that version of the branch had some changes in `desktop/package.json` or `yarn.lock`, either of which would trigger a full artifact build. However the important thing is the artifact set looks correct when isolated to a single architecture. Hope that makes sense.

As a sanity check, if you like you can [temporarily set `ELECTRON_BUILDER_ARGS` to an empty string](https://github.com/Automattic/wp-calypso/blob/cd6333a03b8a04e88b43c2eca1f83878eed87868/desktop/bin/build-mac-ci.js#L10) and `isReleaseBuild` to `true` in the mac release script to compare what artifacts will be persisted at release time (in this case, the artifact set should look more like the "full" set you noted). The electron builder args are set [here](https://github.com/Automattic/wp-calypso/blob/cd6333a03b8a04e88b43c2eca1f83878eed87868/.circleci/config.yml#L199C66-L199C66), and a full build is invoked when dependencies change or the version number is bumped. (Therefore most jobs in trunk will have the smaller artifact set, except when dependencies change, which should be much less frequent.)

Thanks for being so thorough!

**Comment by worldomonation** [comment]
Confirmed that if I set `ELECTRON_BUILDER_ARGS = ''` and `isReleaseBuild = true` in the script, the longer artifact list is produced:
image

With that, I think it's good to go. I will merge this after dropping the experimental changes.

**Pull request merged**
